speaker/bgp_controller.go

```diff
@@ -265,6 +275,30 @@ func (c *bgpController) syncPeers(l log.Logger) error {
 	return nil
 }
 
+func passwordForSession(cfg *config.Peer, bgpType bgpImplementation, secret SecretHandling) (string, v1.SecretReference) {
+	if cfg.SecretPassword != "" && cfg.Password != "" {
+		panic(fmt.Sprintf("non empty password and secret password for peer %s", cfg.Name))
+	}
+	plainTextPassword := cfg.Password
+	if cfg.SecretPassword != "" {
+		plainTextPassword = cfg.SecretPassword
+	}
+	switch bgpType {
+	case bgpNative:
+		return plainTextPassword, v1.SecretReference{}
+	case bgpFrr:
+		return plainTextPassword, v1.SecretReference{}
+	case bgpFrrK8s:
+		// in case of passthrough, we don't propagate the converted password,
+		// so it's either the password as plain text or the secret
+		if secret == SecretPassThrough {
+			return cfg.Password, cfg.PasswordRef
+		}
+		return plainTextPassword, v1.SecretReference{}
+	}
+	return "", v1.SecretReference{}
```

Comment:
imo it feels a bit odd to have this caring about the bgp implementation.
without the new `SecretPassword` the `Password` field acted as "plain text password" (either if it came from the crd directly or a secret)
would it work if we had this function being:
`if passthrough` => `return cfg.Password, cfg.PasswordRef`
`if convert` => `return cfg.Password, v1.SecretReference{}`
if so I think it is cleaner (letting the bgp implementation to decide)
I think this makes sense as the frr-k8s controller itself has the logic of when should it use the ref vs plaintext:
```go
			password := ""
			if reflect.DeepEqual(s.PasswordRef, corev1.SecretReference{}) {
				password = s.Password
			}
```
wdyt?

Reply:
The choice here was to be explicit from above (ie when I call the bgp implementation) or doing some extra logic under the api surface.
I could've propagated the passthrough information, but then we would've been passing two password information without knowing (or, relying on) some hidden behavior down the carpet.

Propagating that to frr-k8s itself is even worse imo, because again, we fill two fields relying on the behavior implemented in a different process.

Here each implementation is doing what we are telling it to do from above, without doing extra hidden things like using only the password or the secret.

Reply:
I understand the rationale (not sure I like it better, but this is nothing critical). If we're going that way we can remove that code snippet I attached?
`Propagating that to frr-k8s itself is even worse imo`
I agree, I was talking about the metallb frr-k8s-controller (where the code snippet is from) 😅

Reply:
ah lol! Sorry, I thought that was frr-k8s.

Yep, we can remove / log / panic if both are set down there.